src/expr.rs

```diff
@@ -1716,12 +1716,19 @@ pub(crate) fn rewrite_field(
         let overhead = name.len() + separator.len();
         let expr_shape = shape.offset_left(overhead)?;
         let expr = field.expr.rewrite(context, expr_shape);
-
         match expr {
-            Some(ref e) if e.as_str() == name && context.config.use_field_init_shorthand() => {
-                Some(attrs_str + name)
+            Some(ref e) => {
+                let is_lit = if let ast::ExprKind::Lit(_) = (*field.expr).kind {
+                    true
+                } else {
+                    false
+                };
+                if !is_lit && e.as_str() == name && context.config.use_field_init_shorthand() {
+                    Some(attrs_str + name)
+                } else {
+                    Some(format!("{}{}{}{}", attrs_str, name, separator, e))
+                }
             }
```

Comment:
Here's a suggestion to minimize the diff. We can check if the `expr.kind` is an `ast::ExprKind::Lit(_)` before we get to the match expression. Then we can just update the first match arm to include the additional check instead of pulling in the logic from the first and second match arms into one arm.

```rust
        let is_lit = matches!(expr.kind, ast::ExprKind::Lit(_));
        match expr {
            Some(ref e) if !is_lit && e.as_str() == name && context.config.use_field_init_shorthand() => {
                Some(attrs_str + name)
            }
            Some(e) => Some(format!("{}{}{}{}", attrs_str, name, separator, e)),
        }
        ...
```